### ISSUE TITLE
CORS Configuration

### ISSUE THREAD

**Issue by seryckd-jda** [hawtio/hawtio#2368]
Hawtio is being flagged during our penetration testing as having an overly permissive CORS setup; as it always returns Access-Control-Allow-Origin: *.

There's an allowAny() method in the CORSFilter that is marked with a comment - 'TODO - allow configuration'.

Is that on a roadmap anywhere?  Or I could I look at adding a configuration property, perhaps hawtio.corsAllowAny, to control this?

**Comment by tadayosi** [comment]
@seryckd-jda Yes, currently `CORSFilter` always returns `*` but it's mostly reflected from the fact that Jolokia by default allows any origins.
https://ro14nd.de/Jolokia-CORS

I believe the Jolokia endpoint is the only part in hawtio that exposes API externally, so `CORSFilter` should behave in accordance with Jolokia's CORS policy. Jolokia can be configured through `jolokia-access.xml`. AFAIK it's difficult for hawtio to provide a sensible restrictive default setting for CORS, so I'm not sure if there is any room for improvements in the hawtio codebase at this moment regarding this.

Please see my blog post regarding how to secure hawtio for more details:
https://medium.com/@tadayoshi.sato/securing-hawtio-f5fbfd5afcf0

**Comment by seryckd-jda** [comment]
Ok, that makes sense for the CORS configuration, thank you for answering.

A penetration test of hawtio also flagged up the CSRF issues that you talk about in your excellent blog article.

Question: Your blog entry shows how to lock down Jolokia endpoints, but how about the other servlets in Hawtio that handle doPost()? Wouldn't they need CSRF protection as well?

For instance,
https://github.com/hawtio/hawtio/blob/master/hawtio-system/src/main/java/io/hawt/web/LoginServlet.java
https://github.com/hawtio/hawtio/blob/master/hawtio-system/src/main/java/io/hawt/web/GitServlet.java
https://github.com/hawtio/hawtio/blob/master/hawtio-system/src/main/java/io/hawt/web/UploadServlet.java

**Comment by tadayosi** [comment]
Thanks @seryckd-jda for your feedback. Good catch. While no such vulnerabilities are reported yet other than the Jolokia servlet, I agree that we should protect those hawtio servlets against CSRF as well. (Actually, I'm now working on a different security issue with UploadServlet, but it could be related to CSRF eventually.)

At this moment, unlike Jolokia servlet I have a feeling that those other servlets may be protected by some classical token-based approach. Let me explore if such protection can be easily implemented.

If token-based approaches don't work, then we might end up needing to have yet another whitelist for those hawtio servlets, but I'd like to avoid forcing users to maintain duplicate whitelists (Jolokia & hawtio).

**Comment by seryckd-jda** [comment]
I'll not take credit for spotting this, we had a security audit that identified possible CSRF vulnerabilities and then ran a fork of Hawtio against [Kiuwan](https://www.kiuwan.com/) which identified the servlets that implemented doPost(). So I do not think there is an existing token implementation, but I may have missed it.

What do you think of the [OWASP CSRFGuard Project](https://www.owasp.org/index.php/Category:OWASP_CSRFGuard_Project)? It's a library that provides an implementation of CSRF tokens through a filter on the server and DOM manipulation on the client. 

I was wondering if it was worth looking into or whether it would be a waste of time if there was no interest from the Hawtio community.

**Comment by tadayosi** [comment]
OWASP CSRFGuard looks quite promising. It's good partly because it's BSD-licensed so hawtio (APL-2.0) can use it.

I'm sure security is a primary concern for the hawtio community, which is used in many enterprises. Security reports like this are always welcomed.

**Comment by tadayosi** [comment]
And contributions are always welcomed too :-)

**Comment by seryckd-jda** [comment]
Ok, I can take a hint :)

There's enough information here to close on the CORS issue, so I will close this.

If I make any progress on OWASP CSRFGuard I will open a new issue.

@tadayosi thanks for you help

**Comment by hoomanb1** [comment]
Just another comment to this discussion so I'm not sure about the feasibility of OWASP CSRFGuard for integration into hawtio as this might need quite a bit of refactoring from the hawtio side to enable dynamic token injection into hawtio pages, I believe the easiest approach to lock down hawtio against XSRF could possibly be a proxy servlet to inspect **origin, X-Forwarded-For, and  Referer** headers for verification. IMO this could be the easiest way of dealing with XSRF without too much regression.

**Comment by hoomanb1** [comment]
... this might not be a 100% solution as some of the above header could still be spoofed but at least it makes it a little harder for a successful exploit.

**Comment by tadayosi** [comment]
@hoomanb1 Thanks for the comment. 

One thing to note is that this is not only the CSRF issue for the proxy servlet but also for all other servlets in hawito. And I think one problem with the proposed approach is, what list of hosts do you actually allow for `Origin`, `X-Forwarded-For`, and `Referer` headers?  I think for most of the servlets the same origin policy may be sufficient but I don't know if the whitelist can be automatically constructed.

Should we provide a generic servlet filter that is configurable by users and checks the `Origin`, `X-Forwarded-For`, and `Referer` headers against the configured list?

**Comment by seryckd-jda** [comment]
@hoomanb1 Just so you are aware, the CSRFGuard project can use DOM manipulation to automatically inject tokens into the server requests. ([CSRF Token Injection](https://www.owasp.org/index.php/CSRFGuard_3_Token_Injection)).  At least, that is the promise.

**Comment by seryckd-jda** [comment]
@hoomanb1, @tadayosi Just an update on CSRFGuard.  I did spend some time implementing this library in Hawtio but the results weren't as satisfactory as I would have liked.

It was easy to integrate the CSRFGuard library into hawtio-system and hawtio-web and get CSRF tokens enabled for the Hawtio plugins. The library uses JS to inject the CSRF token into the DOM after it has been loaded. It can be configured to add hidden fields to Forms or parameters to URIs.

Upon logout or a session timeout the CSRF token would be refreshed in the server but not on the client. Only a browser refresh would force the new token to be retrieved. The root cause of this is that the CSRFGuard library relies upon a page load to refresh the token, and since Hawtio is based upon AngularJS it only loads the page once. I have since learned that the CSRFGuard project is not suitable for Single Page Applications for this reason.

To work around that I added a Core.preLoginTask queue so that the CSRFGuard library could add a task to refresh the token. This worked fine, but then showed up an issue with a false positive CSRF attack at logout. The charts.ts module schedules a Jolokia call and it was possible for the call to be made after logout when the CSRF token was cleared; which showed up as a CSRF warning in the log file (not desirable!). I worked around that by introducing an isLoggedIn boolean to the root scope.

I did try to test Keycloak but ran into an issue related to running Keycloak on the same domain as Hawtio (localhost). CSRFGuard overrides XHR calls to always add a CSRF token to calls made to the same domain. Keycloak has a strict CORS implementation and rejects custom headers. CSRFGuard javascript would have to modified to use port in its definition of same domain so the custom header is not added to Keycloak calls.

Also, multiple Hawtio instances would all need to share the same CSRF token. This may be done with session sharing but I don't have much experience with that.

In summary, it does not look like CSRFGuard is a good fit for Hawtio.
